### PR TITLE
feat(meetings): Event for meeting state change 

### PR DESCRIPTION
**Pull request by arun3528**
# Pull Request Template

## Description
Added Event for meeting state change and log reason as part of a meeting leave 

Also does minor fix to add the conversation Url to the meeting object if missing 

Add leave reason and log it on the console so that the user know was it a client side or server side leave as well as to send metrics

### CODE REVIEW

**Comment by adamweeks** [comment]
> #neglect the calendar plugin changes as its in a different PR

Not sure I understand what you mean by this. Is this meant to be implemented after another PR? Do you have a link to the PR?

If this was meant to be taken as one PR, could you please separate out the calendar and meetings changes into different commits?

**Comment by arun3528** [comment]
@adamweeks  here is the link for the PR for calendar https://github.com/webex/webex-js-sdk/pull/1534

This needs to go in before this meetings event PR

**Comment by adamweeks** [comment]
> @adamweeks here is the link for the PR for calendar #1534
> 
> This needs to go in before this meetings event PR

Thanks @arun3528. In the future, please make sure to let the reviewer know this in the PR description to avoid confusion.

File: packages/node_modules/@webex/internal-plugin-calendar/test/integration/spec/calendar.js [hunk 1]
```diff
@@ -89,7 +89,7 @@ function postToWhistler(webex, type, payload) {
     }));
 }
 
-describe.skip('plugin-calendar', () => {
+describe('plugin-calendar', () => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-calendar/test/integration/spec/calendar.js]: Are these tests working now? Why are they no longer skipped and part of this meetings PR?

File: SCRIPTS.md [hunk 2]
```diff
@@ -78,6 +78,14 @@ Keeps the browser open in debug mode so that you can set break points and reload
 npm test -- --packages @webex/webex-core --browser --karma-debug
 ```
 
+# Run the samples automation tests locally
```
Comment (adamweeks) [inline SCRIPTS.md]: Docs updates should be in a separate commit: `docs(scripts): add samples automation docs`

File: wdio.conf.js [hunk 3]
```diff
@@ -93,7 +93,11 @@ exports.config = {
               'dom.webnotifications.enabled': false,
               'media.getusermedia.screensharing.enabled': true,
               'media.navigator.permission.disabled': true,
-              'media.navigator.streams.fake': true
+              'media.navigator.streams.fake': true,
```
Comment (adamweeks) [inline wdio.conf.js]: These kind of changes should be in their own PR with a description of why they were made.

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 4]
```diff
@@ -793,7 +809,7 @@ export const SDP = {
   MAX_FS_8160: 'max-fs=8160',
   B_LINE: 'b=TIAS',
   CARRIAGE_RETURN: '\r\n',
-  BAD_MEDIA_PORTS: [9, 0]
+  BAD_MEDIA_PORTS: [0]
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js]: What was the cause of this change?

File: packages/node_modules/@webex/internal-plugin-calendar/test/integration/spec/calendar.js (hunk 1 above)
Comment (arun3528) [reply packages/node_modules/@webex/internal-plugin-calendar/test/integration/spec/calendar.js]: pulled the old branch so got messed up

File: packages/node_modules/@webex/plugin-meetings/src/constants.js (hunk 4 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: 9 is actually a valid port . It gets generated at start of the call . Previously it was not . Only 0 is bad port now

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/index.js [hunk 5]
```diff
@@ -320,16 +328,16 @@ export default class LocusInfo extends EventsScope {
...
       }
     }
     else if (this.parsedLocus.fullState.type === _MEETING_) {
-      if (this.fullState && this.fullState.state === LOCUS.STATE.INACTIVE ||
-         this.fullState.state === LOCUS.STATE.TERMINATING) {
-        // We are destroying the meeting here . There might be instance where we need to
-        // change the state and not remove the meeting
+      if (this.fullState && (this.fullState.state === LOCUS.STATE.INACTIVE || this.fullState.state === LOCUS.STATE.TERMINATING)) {
+        // TODO: update the meeting state
+        // this.updateMeeting({state: 'ENDED'});
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/index.js]: What is the purpose of this commented code?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 6]
```diff
@@ -93,15 +94,26 @@ import InMeetingActions from './in-meeting-actions';
  * @property {Boolean} sendShare
  */
 
+/**
+  * Media Ready Event
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: This is not a media ready event

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 7]
```diff
@@ -93,15 +94,26 @@ import InMeetingActions from './in-meeting-actions';
...
+  * Emitted when ever there is a meeting state change
+  * @event meeting:stateChange
+  * @instance
+  * @type {Object}
+  * @property {string} currentState previous state of the meeting
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]:
```suggestion
  * @property {String} currentState previous state of the meeting
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 8]
```diff
@@ -93,15 +94,26 @@ import InMeetingActions from './in-meeting-actions';
...
+  */
+
 /**
   * Media Ready Event
   * Emitted when a stream is ready to be rendered
   * @event media:ready
   * @instance
   * @type {Object}
-  * @property {MediaStream} stream the media stream
+  * @property {string} stream the media stream
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: The stream is a string? Also, please uppercase `String`

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 9]
```diff
@@ -934,13 +960,14 @@ export default class Meeting extends StatelessWebexPlugin {
...
 
-        this.leave().then(() => {
+        this.leave({reason: payload.reason}).then(() => {
           LoggerProxy.logger.warn('meeting/index->setUpLocusInfoMeetingListener:DESTROY_MEETING#The meeting has been left, but has not been destroyed, you should see a later event for leave.');
         }).catch((error) => {
           LoggerProxy.logger.error(`meeting/index->setUpLocusInfoMeetingListener:DESTROY_MEETING#Issue with leave for meeting, meeting still in collection: ${this.meeting}, error: ${error}`);
         });
       }
       else {
+        LoggerProxy.logger.info('MEETIGN_REMOVED_REASON', payload.reason);
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]:
```suggestion
        LoggerProxy.logger.info('MEETING_REMOVED_REASON', payload.reason);
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 10]
```diff
@@ -1812,13 +1839,11 @@ export default class Meeting extends StatelessWebexPlugin {
...
             localShare,
             localStream
-          }).then((mediaResponse) => {
-            Promise.resolve({
-              join: joinResponse,
-              media: mediaResponse,
-              local: [localStream, localShare]
-            });
-          })))
+          }).then((mediaResponse) => Promise.resolve({
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: No need to `Promise.resolve` since this is in a promise chain. Simply return the object:

```js
          }).then((mediaResponse) => ({
             join: joinResponse,
             media: mediaResponse,
             local: [localStream, localShare]
```

**Comment by adamweeks** [comment]
Your integration tests are failing with:

```
Successful 1:1 meeting (including Guest)
        ✖ leave on the meeting object
          Firefox 73.0.0 (Mac OS X 10.13.0)
          Chrome 80.0.3987 (Mac OS X 10.13.6)
        TypeError: Cannot read property 'leave' of null
            at Context.<anonymous> (/tmp/packages/node_modules/@webex/plugin-meetings/test/integration/spec/journey.js:527:46 <- /tmp/ba04e6fd83e8358d233d8a1bf2d29279.browserify.js:339113:86)
```

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/index.js (hunk 5 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/locus-info/index.js]: dont think its needed as we. destroy the meeting object anyways

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 8 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: my bad it should be mediaStream

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 11]
```diff
@@ -375,6 +376,20 @@ export const EVENT_TYPES = {
   ERROR: 'error'
 };
 
+export const MEETING_REMOVED_REASON = {
+  SELF_REMOVED: 'SELF_REMOVED', // server or host removed you from the meeting
+  FULLSTATE_REMOVED: 'FULLSTATE_REMOVED', // meeting got dropped ?
+  MEETING_INACTIVE_TERMINATING: 'MEETING_INACTIVE_TERMINATING',
```
Comment (robstax) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js]: comments on what all of these and `call` vs `meeting` would be appreciated.
